app/components/SearchPopover.tsx

```diff
@@ -0,0 +1,164 @@
+      if (query?.length > 0) {
+        return await documents.search(query, { shareId, ...options });
+      }
+      return undefined;
+    },
+    [documents, shareId]
+  );
+
+  const handleSearch = React.useMemo(
+    () =>
+      debounce(async (event: React.ChangeEvent<HTMLInputElement>) => {
+        const { value } = event.target;
+
+        if (value.length) {
+          popover.show();
+          setQuery(event.target.value.trim());
+        } else {
+          popover.hide();
+        }
+      }, 400),
+    [popover]
+  );
+
+  const searchInputRef = popover.unstable_referenceRef;
+  const firstSearchItem = React.useRef<HTMLAnchorElement>(null);
+
+  const handleEscapeList = React.useCallback(
+    () => searchInputRef?.current?.focus(),
+    [searchInputRef]
+  );
+
+  const handleKeyDown = (ev: React.KeyboardEvent<HTMLInputElement>) => {
+    if (ev.key === "Enter") {
+      if (searchResults?.length) {
+        popover.show();
+      }
+    }
+
+    if (ev.key === "ArrowDown" && !ev.shiftKey) {
+      if (searchResults?.length) {
+        if (ev.currentTarget.value.length === ev.currentTarget.selectionStart) {
+          popover.show();
+        }
+        firstSearchItem.current?.focus();
+      }
+    }
+
+    if (ev.key === "ArrowUp") {
+      if (popover.visible) {
+        popover.hide();
+        ev.preventDefault();
+      }
+
+      if (ev.currentTarget.value) {
+        if (ev.currentTarget.selectionEnd === 0) {
+          ev.currentTarget.selectionStart = 0;
+          ev.currentTarget.selectionEnd = ev.currentTarget.value.length;
+          ev.preventDefault();
+        }
+      }
+    }
+
+    if (ev.key === "Escape") {
+      if (popover.visible) {
+        popover.hide();
+        ev.preventDefault();
+      }
+    }
+  };
+
+  return (
+    <>
+      <PopoverDisclosure {...popover}>
+        {(props) => {
+          // props assumes the disclosure is a button, but we want a type-ahead
+          // so we take the aria props, and ref and ignore the event handlers
+          return (
+            <InputSearch
+              aria-controls={props["aria-controls"]}
+              aria-expanded={props["aria-expanded"]}
+              aria-haspopup={props["aria-haspopup"]}
+              ref={props.ref}
+              onChange={handleSearch}
+              onKeyDown={handleKeyDown}
+            />
+          );
+        }}
+      </PopoverDisclosure>
+
+      <Popover
+        {...popover}
+        aria-label={t("Results")}
+        unstable_autoFocusOnShow={false}
+        style={{ zIndex: theme.depths.sidebar + 1 }}
+        shrink
+      >
+        <PaginatedList
+          options={{ query, minWords: 10, maxWords: 11 }}
+          items={searchResults}
+          fetch={performSearch}
+          onEscape={handleEscapeList}
+          empty={
+            <NoResults>{t("No results for {{query}}", { query })}</NoResults>
+          }
+          loading={<PlaceholderList count={3} header={{ height: 20 }} />}
```

Comment:
Is this amount of flashing expected? We need to improve it somehow as it's crazy distracting

https://user-images.githubusercontent.com/380914/161211773-f3d4a2f8-c1ed-4fd0-aba4-59aa106a29ee.mp4

Reply:
Yeah it's a lot — one of the remaining extra todos is to remove the intermittent loading states. The implementation now is "out of the box" but clearly the update is necessary for this to ship.